### ISSUE TITLE
Is there a race condition when allocating stock?

### ISSUE THREAD

**Issue by anentropic** [django-oscar/django-oscar#1811]
there are several methods in the `partner.StockRecord` model that look a bit like this:

```
def allocate(self, amount, reference='', status=''):
        """
        Convenience method for ring-fencing money against this source
        """
        self.amount_allocated += amount
        self.save()
```

I didn't dig too deep but I didn't find any use of `select_for_update` in the source so I think even knowing that everything's wrapped in a per-request transaction there's a possible race condition here?

Should the methods do something like this to be safer?

```
def allocate(self, amount, reference='', status=''):
        """
        Convenience method for ring-fencing money against this source
        """
        self.__class__.objects.select_for_update().filter(pk=self.pk).update(
            amount_allocated=F('amount_allocated') + amount)
```

**Comment by john-parton** [comment]
As far as I can tell, `select_for_update` isn't used any where in the django-oscar code base at all.

This one might be better solved with a `transaction.atomic` decorator, though.

```python
@transaction.atomic
def allocate(self, amount, reference='', status=''):
        """
        Convenience method for ring-fencing money against this source
        """
        self.amount_allocated += amount
        self.save()
```

Could you please write a test case where the current implementation fails due to a race condition?

**Comment by anentropic** [comment]
`atomic` wouldn't solve the potential race condition in your example, as the instance has already been loaded from the db, so the value of `self.amount_allocated` can't be trusted any more.

even if the instance was loaded from the db _within_ the atomic transaction I don't think you can trust it as the transaction does not preclude another user session from changing the value behind your back (assuming default isolation level https://docs.djangoproject.com/en/1.8/ref/databases/#isolation-level http://www.postgresql.org/docs/current/static/transaction-iso.html#XACT-READ-COMMITTED) 

this is why `select_for_update` exists - i.e. in the case where you `get()` the row then later `save()` it, the db would ensure that row is locked

however _I think_ I have gone too far in my example and `select_for_update` is not necessary when doing an operation like `.filter(pk=self.pk).update(amount_allocated=F('amount_allocated') + amount)` since this accomplishes the update within a single query, so it's enough to just be within an atomic transaction

**Comment by anentropic** [comment]
either way, I agree a test case should be made :)

**Comment by john-parton** [comment]
Excellent info on `select_for_update`. 

Looking at it, you might be right that the `select_for_update` is unnecessary, because it doesn't actually read anything, and just does a single `UPDATE` directly.

I am not a testing expert, and I am _certainly_ not an expert at testing race conditions and other concurrency issues. If you [or anyone else] has any insights, I would be more than happy to listen!

Thanks.

**Comment by digicase** [comment]
I've seen a few examples of allocations for stock records not matching orders, since launching a site this week.

By the way, the example code used in this issue is actually from `apps.payment.abstract_models`

**Comment by digicase** [comment]
I'm trying out this for `stockrecord.allocate`

    def allocate(self, quantity):
        """
        Record a stock allocation.

        This normally happens when a product is bought at checkout.  When the
        product is actually shipped, then we 'consume' the allocation.
        """

        self.__class__.objects.select_for_update().filter(pk=self.pk).update(
            num_allocated=Case(
                When(num_allocated__isnull=True,
                     then=(0 + quantity)),
                default=F('num_allocated') + quantity
            ),
        )
    allocate.alters_data = True

I'm not sure if `select_for_update()` is needed here either.

**Comment by digicase** [comment]
By using update(), the post_save signals for the StockRecord are not sent, which is used for Updating low-stock alerts.

**Comment by anentropic** [comment]
coming back on my earlier comments, I think `select_for_update` is not necessary when doing an atomic `update`, you would only use it where you are first selecting, then modifying the instance, then saving

you could manually send those signals for the affected rows, e.g.

```python
from django.db import router
from django.db.models import signals

def allocate(self, quantity):
    """
    Record a stock allocation.

    This normally happens when a product is bought at checkout.  When the
    product is actually shipped, then we 'consume' the allocation.
    """

    self.__class__.objects.filter(pk=self.pk).update(
        num_allocated=Case(
            When(num_allocated__isnull=True,
                 then=(0 + quantity)),
            default=F('num_allocated') + quantity
        ),
    )
    signals.post_save.send(
        sender=self.__class__,
        instance=self,
        created=False,
        raw=False,
        using=router.db_for_write(self.__class__, instance=self)
    )
allocate.alters_data = True
```

**Comment by mvantellingen** [comment]
@anentropic could you check @2281 for me?

**Comment by anentropic** [comment]
looks ok to me

**Comment by sasha0** [comment]
Fixed in #2281.
